s2spy/_resample.py

```diff
@@ -157,25 +157,51 @@ def resample_dataset(calendar, input_data: xr.Dataset) -> xr.Dataset:
     data = data.to_dataset()
     data = data.stack(anch_int=("anchor_year", "i_interval"))
 
-    means_matrix = xr.DataArray.from_dict(
-        {
-            "coords": {
-                "time": {"dims": "time", "data": input_data["time"].values},
-            },
-            "dims": ("anch_int", "time"),
-            "data": create_means_matrix(
-                data["interval"].values, input_data["time"].values
-            ),
-        }
+    # Separate data with time dims (should be resampled), from data without time dims
+    #   (which does not need resampling). Otherwise stacking ALL dims together will
+    #   cause the stacked dimension become needlessly large, making resampling slow.
+    input_data_time = input_data[
+        [var for var in input_data.data_vars if "time" in input_data[var].dims]
+    ]
+    input_data_nontime = input_data[
+        [var for var in input_data.data_vars if "time" not in input_data[var].dims]
+    ]
+
+    stacking_dims = list(input_data_time.dims.keys())
+    stacking_dims.remove("time")
+    if stacking_dims:  # There might not be extra dims to stack!
+        input_data_time = input_data_time.stack(allstack=stacking_dims)
+
+    da_coords = {"anch_int": data["anch_int"]}
+    if stacking_dims:
+        da_coords["allstack"] = input_data_time["allstack"]
+
+    contains_matrix = contains(
+        pd.IntervalIndex(data["interval"].values), input_data_time["time"].values
     )
 
-    resampled_vars = [xr.DataArray] * len(input_data.data_vars)
-    for i, var in enumerate(input_data.data_vars):
-        resampled_vars[i] = xr.dot(input_data[var], means_matrix, dims=["time"]).rename(
-            var
-        )
+    resampled_vars = [xr.DataArray] * len(input_data_time.data_vars)
+    for i, var in enumerate(input_data_time.data_vars):
+        size_input = input_data_time[var].values.shape[1] if stacking_dims else 1
+        resampled_data = np.zeros((size_input, contains_matrix.shape[0]))
+
+        # Note: while looping, contains_matrix will have a size of
+        # (n_intervals * n_anchor_years), making it the most reliably small dim.
+        for j, row in enumerate(contains_matrix):
+            # Note: next line allows for np.median / np.max etc. To be implemented later
+            resampled_data[:, j] = np.mean(input_data_time[var].values[row], axis=0)
```

Comment:
Now you're extracting `input_data_time[var].values` `j` times. Would it be faster still if you factor this out of the loop? 

Also: have you experimented with the loop order? Intuitively current order makes sense, but I'd need to refresh my memory on loop optimization to be sure.

Reply:
I tried your suggestion, it makes it very slightly faster (~1%).

I did now realize I should have rewritten it to `resampled_data[j, :]` as numpy is row-major. Refactoring this resulted in a >10% speed increase.

About the loop order: we are currently looping through each row of the "contains_matrix"; which represents each `pd.Interval` of the calendar. This will generally be the smallest dimension in bigger datasets (as it doesn't care about extra dims such as lat, lon, etc.). More importantly, this gives us the data to resample, as I don't think we could even easily loop over the columns of `contains_matrix` for this.